data/src/test/java/org/apache/iceberg/data/DeleteReadTests.java

```diff
@@ -586,6 +595,75 @@ public void testEqualityDeleteByNull() throws IOException {
     checkDeleteCount(1L);
   }
 
+  /** Regression test for https://github.com/apache/iceberg/issues/11239 */
+  @TestTemplate
+  public void testEqualityDeleteBinaryColumn() throws IOException {
+    testEqualityDeleteWithColumn(
+        "binaryData", (i) -> ByteBuffer.wrap(("binaryData_" + i).getBytes()));
+  }
+
+  @TestTemplate
+  public void testEqualityDeleteStructColumn() throws IOException {
+    testEqualityDeleteWithColumn(
+        "structData",
+        (i) -> {
+          Record structRecord =
+              GenericRecord.create(table.schema().findType("structData").asStructType());
+          structRecord.setField("structInnerData", "structInnerData_" + i);
+          return structRecord;
+        });
+  }
+
+  private <T> void testEqualityDeleteWithColumn(
+      String targetFieldName, Function<Integer, T> targetFieldValueSupplier) throws IOException {
```

Comment:
Let's avoid a functional interface if we can. Seems like the only thing we are actually avoiding is defining dataDeletes? 

I would suggest refactoring this so that instead to either look like the testEqualityDateDeletes or generating another test method
testEqualityDeletes(List<record> data, List<Record> equalityDeletes, List<Record> expected)?

Something like that